libtenzir/include/tenzir/multi_series.hpp

```diff
@@ -0,0 +1,140 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2024 The Tenzir Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#pragma once
+
+#include "tenzir/detail/function.hpp"
+#include "tenzir/series.hpp"
+
+namespace tenzir {
+
+/// A potentially heterogenous series type.
+class multi_series {
+public:
+  template <class Ty>
+  explicit(false) multi_series(basic_series<Ty> s) {
+    parts_.push_back(std::move(s));
+  }
+
+  explicit multi_series(std::vector<series> parts) : parts_{std::move(parts)} {
+  }
+
+  auto length() const -> int64_t {
+    auto result = int64_t{0};
+    for (auto& part : parts_) {
+      result += part.length();
+    }
+    return result;
+  }
+
+  auto value_at(int64_t row) const -> data_view {
+    auto [part, part_row] = get_index(row);
+    return tenzir::value_at(part.get().type, *part.get().array, part_row);
+  }
+
+  auto is_null(int64_t row) const -> bool {
+    auto [part, part_row] = get_index(row);
+    return part.get().array->IsNull(part_row);
+  }
+
+  auto part(size_t idx) -> series& {
+    TENZIR_ASSERT(idx < parts_.size());
+    return parts_[idx];
+  }
+
+  auto part(size_t idx) const -> const series& {
+    TENZIR_ASSERT(idx < parts_.size());
+    return parts_[idx];
+  }
+
+  auto begin() {
+    return parts_.begin();
+  }
+
+  auto begin() const {
+    return parts_.begin();
+  }
+
+  auto end() {
+    return parts_.end();
+  }
+
+  auto end() const {
+    return parts_.end();
+  }
+
+  auto parts() -> std::span<series> {
+    return parts_;
+  }
+
+  auto parts() const -> std::span<const series> {
+    return parts_;
+  }
+
+  // TODO: Lifetime?
+  auto values() const -> generator<data_view> {
+    for (auto& part : parts_) {
+      for (auto value : part.values()) {
+        co_yield std::move(value);
+      }
+    }
+  }
+
+  auto null_count() const -> int64_t {
+    auto result = int64_t{0};
+    for (auto& part : parts_) {
+      result += part.array->null_count();
+    }
+    return result;
+  }
+
+private:
+  auto get_index(int64_t row) const
+    -> std::pair<std::reference_wrapper<const series>, int64_t> {
+    for (auto& part : parts_) {
+      if (row < part.length()) {
+        return {part, row};
+      }
+      row -= part.length();
+    }
+    TENZIR_UNREACHABLE();
+  }
+
+  std::vector<series> parts_;
+};
+
+// TODO: Lifetime??
```

Comment:
This looks like a TODO that should be addressed.

Reply:
I tried out a few things like marker types, but ultimately this is just down to the lack of lifetime annotations. I just removed the comment.

Reply:
Thanks for the explanation. Let's keep it as-is then.